source/common/router/scoped_rds.h

```diff
@@ -119,15 +132,49 @@ class ScopedRdsConfigSubscription
         ScopedRdsConfigSubscription& parent, std::string scope_name,
         envoy::extensions::filters::network::http_connection_manager::v3::Rds& rds,
         Init::Manager& init_manager);
-    ~RdsRouteConfigProviderHelper() { rds_update_callback_handle_->remove(); }
+
+    RdsRouteConfigProviderHelper(ScopedRdsConfigSubscription& parent, std::string scope_name);
+
+    ~RdsRouteConfigProviderHelper() {
+      // Only remove the rds update when the rds provider has been initialized.
+      if (route_provider_) {
+        rds_update_callback_handle_->remove();
+        parent_.stats_.active_scopes_.dec();
+      }
+      if (on_demand_) {
+        parent_.stats_.on_demand_scopes_.dec();
+      }
+    }
     ConfigConstSharedPtr routeConfig() { return route_provider_->config(); }
 
+    // When on demand callback is received from main thread, there are 4 cases.
+    // 1. Scope is not found, post a scope not found callback back to worker thread.
+    // 2. Scope is found but route provider has not been initialized, create route provider.
+    // 3. After route provider has been initialized, if RouteConfiguration has been fetched,
+    // post scope found callback to worker thread.
+    // 4. After route provider has been initialized, if RouteConfiguration is null,
```

Comment:
"is NullConfigImpl"

Reply:
this is better moved to the code side, it's not a function header comment.

Reply:
Will do.